### PR TITLE
feat: check mesh version if compatible with kiali when startup

### PR DESCRIPTION
**Pull request by Xunzhuo**
Signed-off-by: Xunzhuo <mixdeers@gmail.com>

**Describe the change**

_Please describe what the code change is about._

**Issue reference**

_Please, provide the link to the GitHub issue here._

**Backwards incompatible?**

- Is your change introducing backwards incompatible changes?

_describe the changes if appropriate._

**Documentation**

* Does your contribution contain user-visible changes like e.g. new or changed configuration settings?
If so, please also update [README.adoc](README.adoc)

### CODE REVIEW

File: kiali.go [hunk 1]
```diff
@@ -90,6 +90,21 @@ func main() {
 	status.Put(status.CoreCommitHash, commitHash)
 	status.Put(status.ContainerVersion, determineContainerVersion(version))
 
+	authentication.InitializeAuthenticationController(config.Get().Auth.Strategy)
```
Comment (jmazzitelli) [inline kiali.go:93]: since this is being done here, delete the dup call on 108

File: config/config.go [hunk 2]
```diff
@@ -431,6 +431,29 @@ type HealthConfig struct {
 	Rate []Rate `yaml:"rate,omitempty" json:"rate,omitempty"`
 }
 
+// VersionRange version ranges
+type VersionRange struct {
+	Length      int
+	MeshVersion []string
+	KialiLow    []string
+	KialiHigh   []string
```
Comment (jmazzitelli) [inline config/config.go:439]: who/what/where defines these values?

Comment (Xunzhuo) [reply config/config.go:439]: According to https://kiali.io/docs/installation/installation-guide/prerequisites/#istio-version-compatibility

File: kiali.go (hunk 1 above)
Comment (Xunzhuo) [reply kiali.go]: Done.

File: config/config.go (hunk 2 above)
Comment (jmazzitelli) [reply config/config.go:439]: Yes, but where in the code or config is it set? This is something we are going to have to maintain "somewhere" - where is that matrix coded (i.e. where are the json values coming from that fill in this "VersionRange"?)

Comment (Xunzhuo) [reply config/config.go]: Thanks @jmazzitelli, I have added a yaml to record versions.

File: config/config.go [hunk 3]
```diff
@@ -431,6 +431,38 @@ type HealthConfig struct {
 	Rate []Rate `yaml:"rate,omitempty" json:"rate,omitempty"`
 }
 
+// Versions version ranges
+type Versions struct {
+	MeshName     string `yaml:"meshName"`
+	VersionRange []struct {
+		MeshVersion string `yaml:"meshVersion"`
+		KialiLow    string `yaml:"kialiLow"`
+		KialiHigh   string `yaml:"kialiHigh"`
```
Comment (jmazzitelli) [inline config/config.go:440]: I would prefer this more descriptive. "KialiLow" and "KialiHigh" doesn't really tell you what these are. These should be something like "KialiMinimumVersion" and "KialiMaximumVersion":
```suggestion
		KialiMinimumVersion    string `yaml:"kialiMinimumVersion"`
		KialiMaximumVersion    string `yaml:"kialiMaximumVersion"`
```

File: config/config.go [hunk 4]
```diff
@@ -431,6 +431,38 @@ type HealthConfig struct {
...
+	} `yaml:"versionRange"`
+}
+
+// NewVersions return compatible kiali versions for mesh
+func NewVersions() (*Versions, error) {
```
Comment (jmazzitelli) [inline config/config.go:445]: Failing to check for version compatibility doesn't mean Kiali won't work - it just means it doesn't know (just like today :-).

We should not return an error (or at least we should not abort on any error) here. Just log a Warning message like you do and keep going. Kiali should still run as before.

File: kiali.go [hunk 5]
```diff
@@ -90,6 +90,37 @@ func main() {
 	status.Put(status.CoreCommitHash, commitHash)
 	status.Put(status.ContainerVersion, determineContainerVersion(version))
 
+	// check kiali version compatibility with istio
+	meshName, meshVersion, err := status.GetMeshInfo()
```
Comment (jmazzitelli) [inline kiali.go:94]: This is just a question I have: I haven't looked but - does this status.GetMeshInfo require the user to be authenticated in order to get the MeshInfo data from the server? I assume this will work regardless of the auth strategy being used.

File: version.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# compatible kiali version range with istio version
```
Comment (jmazzitelli) [inline version.yaml:1]: I do not like the name of this file. `version.yaml` implies this file contains the version of Kiali, which is not what it is. Rename this file to something more descriptive of what it is - something like `version-compatibility-matrix.yaml`.

File: kiali.go [hunk 7]
```diff
@@ -90,6 +90,37 @@ func main() {
...
+			log.Info("Mesh Name: %v Mesh Version: %v", meshName, meshVersion)
+			status.Put(status.MeshVersion, meshVersion)
+			status.Put(status.MeshName, meshName)
+			status.Put(status.IsCompatible, "true")
+		} else {
+			log.Warningf("Kiali %v is not compatible with %v %v", version, meshName, meshVersion)
+			status.AddWarningMessages(fmt.Sprintf("Kiali %v is not compatible with %v %v", version, meshName, meshVersion))
+			status.Put(status.IsCompatible, "false")
+		}
+	}
```
Comment (jmazzitelli) [inline kiali.go:122]: I haven't run this to confirm, but I strongly suspect this will emit warnings about incompatibility when Kiali is run in an OpenShift Service Mesh environment, which we do not want. We should enhance this feature being added in this PR by adding to the compatibility matrix all the valid Kiali <-> OSSM versions.

**Review by jmazzitelli — DISMISSED** [review]

File: kiali.go (hunk 7 above)
Comment (Xunzhuo) [reply kiali.go:122]: Sure, when doing this task, I think we can check OSSM version at the same time as well.

File: version.yaml (hunk 6 above)
Comment (Xunzhuo) [reply version.yaml:1]: Sure, I will update this name later.

File: config/config.go (hunk 4 above)
Comment (Xunzhuo) [reply config/config.go:445]: Yes, checking version should not impact Kiali running status. I catch the error for logging warning messages.

File: config/config.go (hunk 3 above)
Comment (Xunzhuo) [reply config/config.go:440]: I will update these names, but I will update related codes for it is using somewhere else.

Comment (Xunzhuo) [reply config/config.go]: Done

File: version.yaml (hunk 6 above)
Comment (Xunzhuo) [reply version.yaml]: Done

File: kiali.go (hunk 5 above)
Comment (Xunzhuo) [reply kiali.go:94]: In my view, querying version info from mesh service does not require to be authenticated.

File: version-compatibility-matrix.yaml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      kialiMinimumVersion: "1.26.0"
+      kialiMaximumVersion: "1.28"
+    - meshVersion: "1.7"
+      SMCPVersion: ""
+      kialiMinimumVersion: "1.22.1"
+      kialiMaximumVersion: "1.25"
+    - meshVersion: "1.6"
+      SMCPVersion: ""
+      kialiMinimumVersion: "1.18.1"
+      kialiMaximumVersion: "1.21"
```
Comment (jshaughn) [inline version-compatibility-matrix.yaml:30]: Just to be complete, I think this should go all the way back to meshVersion 1.0 with a maxKiali of 1.17.    But really none of this matrix will be useful because the check won't apply for kiali versions before the one that merges in this code, unless we do some backports and releases of older Kiali lines.

**Review by jshaughn — CHANGES_REQUESTED** [review]

File: status/versions.go [hunk 9]
```diff
@@ -108,6 +108,121 @@ func validateVersion(requiredVersion string, installedVersion string) bool {
...
+			}
+		}
+	}
+
+	return ok
+}
+
+// checkIstioVersion check istio if its version is compatible with kiali
+func checkIstioVersion(istioVersion string, kialiVersion string) bool {
+	ok := false
```
[the remark this answers is not in the document]
Comment (jmazzitelli) [reply status/versions.go:178]: I haven't looked at what happens when this function returns false, but if we disrupt the user session at all, I think that would be a mistake. I think regardless of whether we determine the current version of Istio is "supported" or not, the user should be allowed to continue as-is normally. We should not log them out, we should not restrict access or functionality at all - they should be able to use Kiali normally as before. This new stuff should only result in a visible "warning" to the user. That warning should say something to the effect of, "Hey, if you see something funny happen, the first thing you should do is upgrade Kiali or Istio because the error is probably due to the fact you are running an obsolete/deprecated/unsupported Kiali compared to the Istio impementation you have installed"

In other words, it shouldn't matter what this function returns - the user should continue normally.

In that case (assuming the above is true - the user session is not affected no matter what this check returns), then I'm of no opinion what this should return if we cannot determine whether kiali supports the installed istio or not.

File: status/versions.go [hunk 10]
```diff
@@ -108,6 +108,121 @@ func validateVersion(requiredVersion string, installedVersion string) bool {
...
+	Versions, err := config.NewVersions()
+
+	if err != nil {
+		log.Warningf("Can not load version file.%v", err)
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply status/versions.go:182]: We should not return here better, for this is kiali.go where impacts the actions of kiali.

[the remark this answers is not in the document]
Comment (Xunzhuo) [reply status/versions.go:182]: I add if .. else .. to check if we should continue in the function.

File: kiali.go [hunk 11]
```diff
@@ -237,3 +222,25 @@ func determineContainerVersion(defaultVersion string) string {
...
+func checkVersionCompatibility(meshName string, meshVersion string) {
+	ok := false
+	if meshName == "Unknown" {
+		log.Warningf("Unknown Istio Implementation")
+		status.Put(status.MeshVersion, meshVersion)
+		status.Put(status.MeshName, meshName)
+		status.AddWarningMessages("Unknown Istio Implementation")
+	} else {
+		ok = status.CheckMeshVersion(meshName, meshVersion, version)
+		if ok {
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply kiali.go]: Done

File: kiali.go [hunk 12]
```diff
@@ -237,3 +222,25 @@ func determineContainerVersion(defaultVersion string) string {
...
+			log.Info("Mesh Name: %v Mesh Version: %v", meshName, meshVersion)
```
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply kiali.go]: Done

File: status/versions.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply status/versions.go:184]: Done and fix also in checkMaistraVersion and checkOSSMVersion

File: version-compatibility-matrix.yaml (hunk 8 above)
Comment (Xunzhuo) [reply version-compatibility-matrix.yaml:30]: But if meshVersion is 1.0, what is the minimum Version of kiali, which is compatible?

File: status/versions.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Xunzhuo) [reply status/versions.go:180]: Actually I followed this rule which this task will not affect kiali normally running status, which means no matter the default ok is true or false, the user should continue normally.

File: version-compatibility-matrix.yaml [hunk 13]
```diff
@@ -0,0 +1,70 @@
+- meshName: "Istio"
+  versionRange: 
+    - meshVersion: "1.12"
+      SMCPVersion: ""
+      kialiMinimumVersion: "1.42.0"
+      kialiMaximumVersion: ""
```
Comment (jshaughn) [inline version-compatibility-matrix.yaml:6]:
```suggestion
    - meshVersion: "1.13"
      SMCPVersion: ""
      kialiMinimumVersion: "1.45.1"
      kialiMaximumVersion: ""
    - meshVersion: "1.12"
      SMCPVersion: ""
      kialiMinimumVersion: "1.42.0"
      kialiMaximumVersion: "1.44.0"
```

File: kiali.go [hunk 14]
```diff
@@ -211,3 +222,25 @@ func determineContainerVersion(defaultVersion string) string {
...
+		status.Put(status.MeshName, meshName)
+		status.AddWarningMessages("Unknown Istio Implementation")
+	} else {
+		if ok := status.CheckMeshVersion(meshName, meshVersion, version); ok {
+			log.Infof("Mesh Name: %v Mesh Version: %v", meshName, meshVersion)
+			status.Put(status.MeshVersion, meshVersion)
+			status.Put(status.MeshName, meshName)
+			status.Put(status.IsCompatible, "true")
+		} else {
+			log.Warningf("Kiali %v is not compatible with %v %v", version, meshName, meshVersion)
```
Comment (jshaughn) [inline kiali.go:241]: This message is fine for OSSM and Maistra, but for upstream Istio it may be too agressive.  We know it's not recommended (because it's not tested) but we may not be sure if it's incompatible.   I would suggest something like:

`log.Warningf("Kiali %v may not be compatible with %v %v, and is not recommended. See kiali.io for version compatibility", version, meshName, meshVersion)`

**Review by jshaughn — CHANGES_REQUESTED** [review]
Looks good.  Only one suggested change.

File: version-compatibility-matrix.yaml (hunk 8 above)
Comment (jmazzitelli) [reply version-compatibility-matrix.yaml:30]: what @jshaughn proposes seems reasonable.

File: kiali.go (hunk 14 above)
Comment (jmazzitelli) [reply kiali.go:241]: Yeah, let's warn them that it MAY NOT be compatible. I think the suggested change is good.

File: version-compatibility-matrix.yaml (hunk 13 above)
Comment (Xunzhuo) [reply version-compatibility-matrix.yaml]: Done.

File: kiali.go (hunk 14 above)
Comment (Xunzhuo) [reply kiali.go]: Done.

**Review by jshaughn — CHANGES_REQUESTED** [review]
Please add the remaining versions to the matrix, as described in the comment.

**Comment by Xunzhuo** [comment]
@jshaughn Done and Thanks. :)

**Review by jshaughn — CHANGES_REQUESTED** [review]
Code looks good, compat matrix looks good, but did you test this using an actual build?  I don't
think `version-compatibility-matrix.yaml` is actually getting included in the container.  Try `make build cluster-push operator-create kiali-create` and I think you will see `WRN Can not load version file: open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory`

**Comment by jmazzitelli** [comment]
> Code looks good, compat matrix looks good, but did you test this using an actual build? I don't think `version-compatibility-matrix.yaml` is actually getting included in the container. Try `make build cluster-push operator-create kiali-create` and I think you will see `WRN Can not load version file: open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory`

I suspect the file needs to be added [here](https://github.com/kiali/kiali/blob/master/deploy/docker/Dockerfile-multi-arch#L23) and [here](https://github.com/kiali/kiali/blob/master/deploy/docker/Dockerfile-ubi8-minimal#L17). Something like:

```
ADD <local path of the matrix yaml> $KIALI_HOME
```

File: Makefile [hunk 15]
```diff
@@ -19,6 +19,9 @@ COMMIT_HASH ?= $(shell git rev-parse HEAD)
 # output in the $UI_SRC_ROOT/build/ subdirectory.
 CONSOLE_LOCAL_DIR ?= ${ROOTDIR}/../../../../../kiali-ui
 
+# The path where the version-compatibility-matrix.yaml has been git cloned.
+VERSION_METRIX_LOCAL_DIR ?= ${ROOTDIR}/version-compatibility-matrix.yaml
```
Comment (jmazzitelli) [inline Makefile:23]: I just realized in several places the word "matrix" is being misspelled - here (and elsewhere) you have "metrix" (with an "e" not an "a") --- the correct spelling is "matrix".

File: make/Makefile.container.mk [hunk 16]
```diff
@@ -9,6 +9,7 @@
 	@mkdir -p ${OUTDIR}/docker
 	@cp -r deploy/docker/* ${OUTDIR}/docker
 	@cp ${GOPATH}/bin/kiali* ${OUTDIR}/docker
+	@cp ${VERSION_METRIX_LOCAL_DIR} ${OUTDIR}/docker
```
Comment (jmazzitelli) [inline make/Makefile.container.mk:12]: misspelled "matrix"

File: version-compatibility-matrix.yaml [hunk 17]
```diff
@@ -0,0 +1,103 @@
+# The Compatible Version Metrix between upstream Istio and Kiali
```
Comment (jmazzitelli) [inline version-compatibility-matrix.yaml:1]: misspelled "matrix"

File: version-compatibility-matrix.yaml [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    - meshVersion: "1.1"
+      SMCPVersion: ""
+      kialiMinimumVersion: "1.17"
+      kialiMaximumVersion: "1.17"
+    - meshVersion: "1.0"
+      SMCPVersion: ""
+      kialiMinimumVersion: "1.17"
+      kialiMaximumVersion: "1.17"
+
+# The Compatible Version Metrix between Maistra and Kiali
```
Comment (jmazzitelli) [inline version-compatibility-matrix.yaml:61]: misspelled "matrix"

File: version-compatibility-matrix.yaml [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    - meshVersion: "2.0"
+      SMCPVersion: "1.1"
+      kialiMinimumVersion: "1.12"
+      kialiMaximumVersion: ""
+    - meshVersion: "1.1"
+      SMCPVersion: "1.1"
+      kialiMinimumVersion: "1.12"
+      kialiMaximumVersion: ""
+
+# The Compatible Version Metrix between OpenShift Service Mesh and Kiali
```
Comment (jmazzitelli) [inline version-compatibility-matrix.yaml:89]: misspelled "matrix"

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
trivial change - correct the misspellings of the word "matrix"

File: Makefile (hunk 15 above)
Comment (Xunzhuo) [reply Makefile]: Done.

File: make/Makefile.container.mk (hunk 16 above)
Comment (Xunzhuo) [reply make/Makefile.container.mk]: Done.

File: version-compatibility-matrix.yaml (hunk 17 above)
Comment (Xunzhuo) [reply version-compatibility-matrix.yaml]: Done.

**Review by jmazzitelli — COMMENTED** [review]
still testing, but so far so good - I see the matrix file has now made it into the image.

1. deploy everything to minikube
```
make -e CLUSTER_TYPE=minikube make build cluster-push operator-create kiali-create
```

2. find the matrix file:
```
$ kubectl exec -it -n istio-system $(kubectl get pod -n istio-system -l app.kubernetes.io/name=kiali -o name) -- ls -l /opt/kiali/version-compatibility-matrix.yaml

-rw-rw-r-- 1 root root 2973 Feb  9 14:26 /opt/kiali/version-compatibility-matrix.yaml
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
@Xunzhuo you need to test this when deploying in a cluster (i.e. unit tests won't be enough to validate).

I'm still getting the error @jshaughn got - the path name looks incorrect:

```
2022-02-09T14:27:20Z WRN Can not load version file open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory
2022-02-09T14:27:20Z WRN Can not load version file: open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory
```

To test, start up minikube, install Istio, and deploy using the make target as I document in my previous comment. We have make targets and hack scripts to help you do all of this, e.g.:

```
$ hack/k8s-minikube.sh start
$ hack/k8s-minikube.sh istio
$ make -e CLUSTER_TYPE=minikube make build cluster-push operator-create kiali-create
```

Those 3 commands will get you a minikube cluster with Kiali installed, at which time you can look at the logs via `kubectl logs` and peek into the image via `kubectl exec`

**Comment by Xunzhuo** [comment]
> A secondary comment - I see the warning twice:
> 
> ```
> 2022-02-09T14:27:20Z WRN Can not load version file open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory
> 2022-02-09T14:27:20Z WRN Can not load version file: open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory
> ```
> 
> which tells me we are checking version compatibility multiple times. We only need to do it once - so just make sure we don't validate more than once. It is also possible that rather than checking twice, we are logging the message twice through the call stack (I say that because the warning message is ever so slightly different - notice the extra `:` (colon) after "file" in the second message). So just make sure we only log this once - no need to repeat the warning.

Actually I do check version once, just a duplicated log, solve this just by moving it.

**Comment by Xunzhuo** [comment]
> @Xunzhuo you need to test this when deploying in a cluster (i.e. unit tests won't be enough to validate).
> 
> 
> 
> I'm still getting the error @jshaughn got - the path name looks incorrect:
> 
> 
> 
> ```
> 
> 2022-02-09T14:27:20Z WRN Can not load version file open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory
> 
> 2022-02-09T14:27:20Z WRN Can not load version file: open /opt/kiali/../version-compatibility-matrix.yaml: no such file or directory
> 
> ```
> 
> 
> 
> To test, start up minikube, install Istio, and deploy using the make target as I document in my previous comment. We have make targets and hack scripts to help you do all of this, e.g.:
> 
> 
> 
> ```
> 
> $ hack/k8s-minikube.sh start
> 
> $ hack/k8s-minikube.sh istio
> 
> $ make -e CLUSTER_TYPE=minikube make build cluster-push operator-create kiali-create
> 
> ```
> 
> 
> 
> Those 3 commands will get you a minikube cluster with Kiali installed, at which time you can look at the logs via `kubectl logs` and peek into the image via `kubectl exec`
> 
> 
> 
> For the record, here's what I see in the /opt/kiali directory inside the image:
> 
> 
> 
> ```
> 
> $ kubectl exec -it -n istio-system $(kubectl get pod -n istio-system -l app.kubernetes.io/name=kiali -o name) -- ls -l /opt/kiali
> 
> total 61468
> 
> drwxrwxr-x 1 kiali kiali     4096 Feb  9 14:26 console
> 
> -rwxrwxr-x 1 root  root  62931354 Feb  9 14:26 kiali
> 
> -rw-rw-r-- 1 root  root      2973 Feb  9 14:26 version-compatibility-matrix.yaml
> 
> ```
> 
> 

Actually I try to run these commands to test, but failed for a couple of times. I tried to set ClusterType to local, but there are still some conflicts with my env. I will try to fix them later, I think there are some bugs in these scripts. But I need to retest to confirm.

**Comment by jmazzitelli** [comment]
We haven't tested the scripts on, say, MacOS. I run them on RHEL 7.x. So it might have to do with whatever operating system you are on. CLUSTER_TYPE=local won't work most likely - that's for special situations (which I forget).

**Comment by jmazzitelli** [comment]
> I tried to run these commands to test, but failed for a couple of times.

You don't have to use these scripts - however you work in your environment is fine. So long as you deploy Kiali in a cluster that also has Istio.

**Comment by Xunzhuo** [comment]
Thanks @jmazzitelli @jshaughn , I tested and deploy kiali into cluster. I think everything is fine now.

image

**Review by jshaughn — DISMISSED** [review]
This is working for me, looks good!

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
@Xunzhuo I got around to doing a more deep-dive review and testing of this. I have some requested changes - but I did them already in a PR to your PR - please see: https://github.com/Xunzhuo/kiali/pull/1

**Comment by jmazzitelli** [comment]
this is what the UI will look like:
image

**Comment by Xunzhuo** [comment]
looks good @jmazzitelli !

**Review by jmazzitelli — APPROVED** [review]
ran this and it works. lgtm
